modules/transaction-payment/src/lib.rs

```diff
@@ -578,59 +642,199 @@ where
 				});
+		if native_is_enough {
+			return;
+		}
 
-		// native is not enough, try swap native to pay fee and gap
-		if !native_is_enough {
-			// add extra gap to keep alive after swap
-			let amount = fee.saturating_add(native_existential_deposit.saturating_sub(total_native));
-			let native_currency_id = T::NativeCurrencyId::get();
-			let default_fee_swap_path_list = T::DefaultFeeSwapPathList::get();
-			let fee_swap_path_list: Vec<Vec<CurrencyId>> =
-				if let Some(trading_path) = AlternativeFeeSwapPath::<T>::get(who) {
-					vec![vec![trading_path.into_inner()], default_fee_swap_path_list].concat()
-				} else {
-					default_fee_swap_path_list
-				};
-
-			for trading_path in fee_swap_path_list {
-				match trading_path.last() {
-					Some(target_currency_id) if *target_currency_id == native_currency_id => {
-						let supply_currency_id = *trading_path.first().expect("these's first guaranteed by match");
-						// calculate the supply limit according to oracle price and the slippage limit,
-						// if oracle price is not avalible, do not limit
-						let max_supply_limit = if let Some(target_price) =
-							T::PriceSource::get_relative_price(*target_currency_id, supply_currency_id)
-						{
-							Ratio::one()
-								.saturating_sub(T::MaxSwapSlippageCompareToOracle::get())
-								.reciprocal()
-								.unwrap_or_else(Ratio::max_value)
-								.saturating_mul_int(target_price.saturating_mul_int(amount))
-						} else {
-							PalletBalanceOf::<T>::max_value()
-						};
+		// native is not enough, try swap native from charge fee pool to pay fee and gap.
+		// make sure add extra gap to keep alive after swap.
+		let amount = fee.saturating_add(native_existential_deposit.saturating_sub(total_native));
+		Self::swap_native_asset(who, amount);
+	}
+
+	/// Iterate order list, if can swap out enough native asset amount with foreign asset, then stop
+	fn swap_native_asset(who: &T::AccountId, amount: Balance) {
+		let native_currency_id = T::NativeCurrencyId::get();
+		for trading_path in Self::get_trading_path(who) {
+			match trading_path.last() {
+				Some(target_currency_id) if *target_currency_id == native_currency_id => {
+					let supply_currency_id = *trading_path.first().expect("these's first guaranteed by match");
+					if Self::swap_from_treasury(who, amount, supply_currency_id).is_ok() {
+						break;
+					}
+				}
+				_ => {}
+			}
+		}
+	}
+
+	/// swap from treasury
+	#[transactional]
+	fn swap_from_treasury(
+		who: &T::AccountId,
+		amount: Balance,
+		supply_currency_id: CurrencyId,
+	) -> sp_std::result::Result<(), DispatchError> {
+		let treasury_account = Self::treasury_sub_account_id(supply_currency_id);
+		let treasury_supply_balance = T::MultiCurrency::free_balance(supply_currency_id, &treasury_account);
+		let mut rate = FeeRateOfToken::<T>::get(supply_currency_id)
+			.ok_or(Error::<T>::AssetNotInFeePool)
+			.unwrap();
```

Comment:
```suggestion
			.unwrap();
```

`unwrap` is not allowed in runtime, return error or use `expect` with proofs.